src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/CaseDocumentService.java

```diff
@@ -0,0 +1,134 @@
+    }
+
+    public CaseDetails stripDocumentHashes(final CaseDetails caseDetails) {
+        final List<JsonNode> documentNodes = caseDocumentUtils.findDocumentNodes(caseDetails.getData());
+
+        return documentNodes.isEmpty() ? caseDetails : removeHashes(caseDetails);
+    }
+
+    public List<DocumentHashToken> extractDocumentHashToken(final Map<String, JsonNode> databaseCaseData,
+                                                            final Map<String, JsonNode> preCallbackCaseData,
+                                                            final Map<String, JsonNode> postCallbackCaseData) {
+
+        final List<Tuple2<String, String>> dbDocs = caseDocumentUtils.findDocumentsHashes(
+            databaseCaseData
+        );
+
+        final List<Tuple2<String, String>> eventDocs = caseDocumentUtils.findDocumentsHashes(
+            preCallbackCaseData
+        );
+
+        final List<Tuple2<String, String>> postCallbackDocs = caseDocumentUtils.findDocumentsHashes(
+            postCallbackCaseData
+        );
+
+        final List<Tuple2<String, String>> preCallbackDocs = CollectionUtils.listsUnion(dbDocs, eventDocs);
+
+        verifyNoTamper(preCallbackDocs, postCallbackDocs);
+
+        final List<DocumentHashToken> documentHashTokens = caseDocumentUtils.buildDocumentHashToken(
+            preCallbackDocs,
+            postCallbackDocs
+        );
+
+        validate(documentHashTokens);
+
+        return documentHashTokens;
+    }
+
+    public List<DocumentHashToken> extractDocumentHashToken(final Map<String, JsonNode> preCallbackCaseData,
+                                                            final Map<String, JsonNode> postCallbackCaseData) {
+
+        return extractDocumentHashToken(emptyMap(), preCallbackCaseData, postCallbackCaseData);
+    }
+
+    public void attachCaseDocuments(final String caseId,
+                                    final String caseTypeId,
+                                    final String jurisdictionId,
+                                    final List<DocumentHashToken> documentHashes) {
+        if (documentHashes.isEmpty()) {
+            return;
+        }
+
+        final CaseDocumentsMetadata documentMetadata = CaseDocumentsMetadata.builder()
+            .caseId(caseId)
+            .caseTypeId(caseTypeId)
+            .jurisdictionId(jurisdictionId)
+            .documentHashToken(documentHashes)
+            .build();
+
+        caseDocumentAmApiClient.applyPatch(documentMetadata);
+    }
+
+    void validate(final List<DocumentHashToken> documentHashes) {
+        if (!applicationParams.isDocumentHashCheckingEnabled()) {
+            return;
+        }
+
+        final List<DocumentHashToken> violatingDocuments = caseDocumentUtils.getViolatingDocuments(
+            documentHashes
+        );
+
+        if (!violatingDocuments.isEmpty()) {
+            throw new ValidationException("Some message");  // TODO: suitable error message
```

Comment:
Add error message like "Document hashTokens are missing for the documents: <document-ids>"